curp/src/server/bg_tasks.rs

```diff
@@ -228,20 +235,26 @@ async fn send_log_until_succeed<C: Command + 'static>(
 }
 
 /// Background `append_entries`, only works for the leader
-async fn bg_heartbeat<C: Command + 'static>(
-    connects: Vec<Arc<Connect>>,
+async fn bg_heartbeat<C: Command + 'static, Conn: ConnectInterface>(
+    connects: Vec<Arc<Conn>>,
     state: Arc<RwLock<State<C>>>,
 ) {
-    let role_trigger = state.read().role_trigger();
+    let (role_trigger, hb_reset_trigger) =
+        state.map_read(|state_r| (state_r.role_trigger(), state_r.heartbeat_reset_trigger()));
     #[allow(clippy::integer_arithmetic)] // tokio internal triggered
     loop {
+        loop {
+            tokio::select! {
+                _ = tokio::time::sleep(HEARTBEAT_INTERVAL) => break,
+                _ = hb_reset_trigger.listen() => {}
```

Comment:
I think "Event" is not the perfect data structure here. Imagining that the system is quite busy, hundreds of "append_entry" are called, we need to enter listen hundreds of times.

Even using "Event", the listen() function only returns the listener, the "await" is required.

Reply:
> Even using "Event", the listen() function only returns the listener, the "await" is required.

`tokio::select!` will call await for us.

> I think "Event" is not the perfect data structure here. Imagining that the system is quite busy, hundreds of "append_entry" are called, we need to enter listen hundreds of times.

The listener returned by `listen()` will be consumed every time we call await to it. So we had to call it every time. I guess the alternative way is to use time to check if a heartbeat is needed. This method would require us to acquire system time when `append_entries` is called. I suspect it would be more efficient.

Reply:
I see, the new "Listener" won't get previous notifications, so it's not a hot point.